lib/spack/spack/detection/path.py

```diff
@@ -94,17 +130,45 @@ def libraries_in_ld_and_system_library_path(
             constructed based on the set of LD_LIBRARY_PATH, LIBRARY_PATH,
             DYLD_LIBRARY_PATH, and DYLD_FALLBACK_LIBRARY_PATH environment
             variables as well as the standard system library paths.
+        path_hints (list): list of paths to be searched. If ``None``, the default
+            system paths are used.
     """
-    default_lib_search_paths = (
-        spack.util.environment.get_path("LD_LIBRARY_PATH")
-        + spack.util.environment.get_path("DYLD_LIBRARY_PATH")
-        + spack.util.environment.get_path("DYLD_FALLBACK_LIBRARY_PATH")
-        + spack.util.ld_so_conf.host_dynamic_linker_search_paths()
-    )
-    path_hints = path_hints if path_hints is not None else default_lib_search_paths
-
-    search_paths = llnl.util.filesystem.search_paths_for_libraries(*path_hints)
-    return path_to_dict(search_paths)
+    if path_hints:
+        search_paths = llnl.util.filesystem.search_paths_for_libraries(*path_hints)
+    else:
+        search_paths = []
+
+        # Environment variables
+        if sys.platform == "darwin":
+            search_paths.extend(environment.get_path("DYLD_LIBRARY_PATH"))
+            search_paths.extend(environment.get_path("DYLD_FALLBACK_LIBRARY_PATH"))
+        elif sys.platform.startswith("linux"):
+            search_paths.extend(environment.get_path("LD_LIBRARY_PATH"))
+
+        # Dynamic linker paths
+        search_paths.extend(spack.util.ld_so_conf.host_dynamic_linker_search_paths())
+
+        # Drop redundant paths
+        search_paths = list(filter(os.path.isdir, search_paths))
+
+    # Make use we don't doubly list /usr/lib and /lib etc
+    search_paths = list(llnl.util.lang.dedupe(search_paths, key=file_identifier))
```

Comment:
Since the file identifier conatins the inode, would this dedupe on e.g. Ubuntu:
```console
$ stat /lib /usr/lib
  File: /lib -> usr/lib
  Size: 7         	Blocks: 0          IO Block: 4096   symbolic link
Device: 10303h/66307d	Inode: 14          Links: 1
Access: (0777/lrwxrwxrwx)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2023-11-01 12:05:01.533555114 +0100
Modify: 2021-08-20 16:17:21.544002711 +0200
Change: 2021-08-20 16:17:21.544002711 +0200
 Birth: -
  File: /usr/lib
  Size: 12288     	Blocks: 24         IO Block: 4096   directory
Device: 10303h/66307d	Inode: 25165829    Links: 154
Access: (0755/drwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2023-11-02 09:20:46.064505013 +0100
Modify: 2023-10-28 13:07:40.883302892 +0200
Change: 2023-10-28 13:07:40.883302892 +0200
 Birth: -
```
? Or am I reading this wrong?

Reply:
It's been a while since I wrote that, but I guess the idea is that `os.stat` follows the symlink, it's not `lstat`

Reply:
Hm, do we actually filter existing paths?

Reply:
Ah yes, filter(os.path.isdir) above should do it...

Reply:
Gotcha, `os.stat` maps to `stat -L`, while `os.lstat` maps to `stat` (after reading docs). Weird on the Python side, but correct here.